### PR TITLE
Sovrn connector UI

### PR DESCRIPTION
**Pull request by allisonking**
Closes https://github.com/ethyca/fides/issues/2458

### Code Changes

There was unfortunately quite a bit of tech debt in this section of code with typescript being ignored and other eslint workarounds. I couldn't resist fixing it up (also because things were confusing me). These were:
* [x] Extracting filling in defaults to a helper function (with tests). Making copies of parameters instead of reassigning them
* [x] Refactoring `skip` logic on querying for connection schema
* [x] Fix bug that made connector types with only one tab not render properly 

Actual feature changes
* [x] Update openapi typescript types to include latest email type endpoints
* [x] Add a new Email type ConnectorParameters.tsx (following the pattern of the other connector types)
  * [x] This uses the same `handleSubmit` code as the Database type connector parameter, so I did some refactoring so they could both use the same logic
* [x] Cypress tests

### Steps to Confirm

This is probably best QA'd alongside https://github.com/ethyca/fides/pull/2582

If you make a separate branch out of this one plus https://github.com/ethyca/fides/pull/2582, then you should be able to go through the whole flow:

* Go through the same steps as in https://github.com/ethyca/fides/pull/2543 _except_:
  * Instead of `Because we can't pass in the ljt_readerID from the frontend yet, cache a fake one in the request runner service...`, do the following:
    * Add a cookie to your page by opening up the inspector in your browser (opt+cmd+i on osx), entering the console, and setting your cookie via `document.cookie = "ljt_readerID=fides_test_reader_id;"
    * Issue a consent request via the UI to opt out of Email Marketing. In the network tab, you should see ljt_readerID show up in the payload for PATCH preferences under browser_identity
  * Instead of `Add a sovrn consent email connector via the API`:
    *  Go through the admin UI and add a sovrn connector! You should be able to add it and also send a test email

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`

### Description Of Changes

_Write some things here about the changes and any potential caveats_

### CODE REVIEW

File: clients/admin-ui/src/features/datastore-connections/add-connection/ConfigureConnector.tsx [hunk 1]
```diff
@@ -123,12 +123,6 @@ const ConfigureConnector: React.FC = () => {
   }, [dispatch]);
 
   useEffect(() => {
-    if (step) {
-      setSelectedItem(
-        connector?.options[Number(step.stepId) - connector.options.length]
-      );
-    }
```
Comment (allisonking) [inline clients/admin-ui/src/features/datastore-connections/add-connection/ConfigureConnector.tsx:130]: I really could not figure out the purpose of this line, and it caused a bug when there's only one tab available, so I got rid of it

File: clients/admin-ui/src/features/datastore-connections/add-connection/ConnectorParameters.tsx [hunk 2]
```diff
@@ -67,18 +72,24 @@ export const ConnectorParameters: React.FC<ConnectorParametersProp> = ({
...
-
-  useEffect(() => {
-    mounted.current = true;
-    if (connectionOption?.type !== SystemType.MANUAL) {
-      setSkip(false);
-    }
-    return () => {
-      mounted.current = false;
-    };
-  }, [connectionOption?.type]);
```
Comment (allisonking) [inline clients/admin-ui/src/features/datastore-connections/add-connection/ConnectorParameters.tsx:81]: this `useEffect` + skip pattern was hard to follow, so I refactored it and it seems to still work well

File: clients/admin-ui/src/features/datastore-connections/add-connection/database/ConnectorParameters.tsx [hunk 3]
```diff
@@ -36,19 +39,14 @@ type ConnectorParametersProps = {
   onTestConnectionClick: (value: any) => void;
 };
 
-export const ConnectorParameters: React.FC<ConnectorParametersProps> = ({
-  data,
+export const useDatabaseConnector = ({
```
Comment (allisonking) [inline clients/admin-ui/src/features/datastore-connections/add-connection/database/ConnectorParameters.tsx:42]: I extracted this to a hook because the email connector uses the same API logic

File: clients/admin-ui/src/features/datastore-connections/add-connection/forms/ConnectorParametersForm.tsx [hunk 4]
```diff
@@ -188,32 +193,16 @@ const ConnectorParametersForm: React.FC<ConnectorParametersFormProps> = ({
...
-        // eslint-disable-next-line no-nested-ternary
-        defaultValues[key[0]] = key[1].default
-          ? key[1].default
-          : key[1].type === "integer"
-          ? 0
-          : "";
-      });
     }
-    return defaultValues;
+    return fillInDefaults(initialValues, data);
```
Comment (allisonking) [inline clients/admin-ui/src/features/datastore-connections/add-connection/forms/ConnectorParametersForm.tsx:205]: basically an effort to remove duplicate code (note the two `Object.entries()...` that are the same above) and also to extract to a function that is tested

**Comment by allisonking** [comment]
Oh actually I forgot I still need to fix the `Awaiting Email Send`

**Comment by allisonking** [comment]
image

Another option was:
image

I thought the first one looked slightly better, but it's easy enough to change (the second option uses `whiteSpace="break-spaces"` and removes the fixed width)

File: clients/admin-ui/src/features/connection-type/types.ts [hunk 5]
```diff
@@ -25,6 +31,15 @@ export type ConnectionTypeSecretSchemaReponse = {
...
   type: string;
+  definitions: {
+    AdvancedSettings: {
+      title: string;
+      type: string;
+      properties: { [key: string]: ConnectionTypeSecretSchemaProperty };
+    };
+    IdentityTypes: IdentityTypes;
+    CookieIds: CookieIds;
+  };
```
Comment (pattisdr) [inline clients/admin-ui/src/features/connection-type/types.ts:42]: API now expects something like 

```
{
    "recipient_email_address": "test@example.com",
    "test_email_address": "test@example.com",
    "advanced_settings": {
        "identity_types": {
            "email": false,
            "phone_number": false,
            "cookie_ids":  ["ljt_readerID"]
        }
    }
}

```

**Review by pattisdr — COMMENTED** [review]
@allisonking could you actually change this so it makes a request that looks something like this?  My latest commit on the backend branch yesterday where I change the expected format expects the advanced settings block.

```
{
    "recipient_email_address": "dawn@ethyca.com",
    "test_email_address": "dawn@ethyca.com",
    "advanced_settings": {
        "identity_types": {
            "email": false,
            "phone_number": false,
            "cookie_ids":  []
        }
    }
}
```

File: clients/admin-ui/src/features/datastore-connections/add-connection/ConfigureConnector.tsx (hunk 1 above)
Comment (pattisdr) [reply clients/admin-ui/src/features/datastore-connections/add-connection/ConfigureConnector.tsx:130]: Bravo tracking this down!

File: clients/admin-ui/cypress/fixtures/connectors/sovrn_secret.json [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+      "properties": {
+        "identity_types": {
+          "default": [],
+          "type": "array",
+          "items": { "$ref": "#/definitions/IdentityTypes" }
+        },
+        "browser_identity_types": {
+          "default": [],
+          "type": "array",
+          "items": { "$ref": "#/definitions/CookieIds" }
```
[the remark this answers is not in the document]
Comment (allisonking) [reply clients/admin-ui/cypress/fixtures/connectors/sovrn_secret.json:45]: thanks for pasting the payload right here—helps a lot!!

File: clients/admin-ui/src/features/connection-type/types.ts (hunk 5 above)
Comment (allisonking) [reply clients/admin-ui/src/features/connection-type/types.ts]: updated!

File: clients/admin-ui/cypress/fixtures/connectors/sovrn_secret.json [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  "type": "object",
+  "properties": {
+    "third_party_vendor_name": {
+      "title": "Third Party Vendor Name",
+      "default": "Sovrn",
+      "type": "string"
+    },
+    "recipient_email_address": {
+      "title": "Recipient Email Address",
+      "type": "string"
```
[the remark this answers is not in the document]
Comment (allisonking) [reply clients/admin-ui/cypress/fixtures/connectors/sovrn_secret.json:13]: okay, I think that's not too important to have in this fixture right now. I can add it down the line once it's in 👍

[the remark this answers is not in the document]
Comment (pattisdr) [reply clients/admin-ui/cypress/fixtures/connectors/sovrn_secret.json:13]: I don't even know if I we should add as a default! I feel like it's too easy to accidentally send test emails to Sovrn directly, like it's something that should only be added in prod.

File: clients/admin-ui/src/features/connection-type/types.ts (hunk 5 above)
Comment (pattisdr) [reply clients/admin-ui/src/features/connection-type/types.ts]: thanks so much for this update !

File: clients/admin-ui/cypress/fixtures/connectors/sovrn_secret.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (allisonking) [reply clients/admin-ui/cypress/fixtures/connectors/sovrn_secret.json:13]: oh yeah that's a good point...

**Review by pattisdr — APPROVED** [review]
While I can't easily review the code, I QA'd this and it looks great to me.  Nice with the one tab so DatasetConfigs aren't created.  Thanks again for stubbing out the API response for now.

You're welcome to merge this into my backend base branch when ready!

**Comment by allisonking** [comment]
Thanks @pattisdr ! requesting @TheAndrewJackson to review the code part. I think you've done the hard part of QA-ing it, so hopefully Andrew can just look through the code (not that that is easy either, sorry for all the changes... it took me a long time to wrap my head around how the connectors stuff worked)
